### PR TITLE
Contact hydration

### PR DESCRIPTION
**Pull request by willdoran**
This pull request makes the following changes:
- Adds contacts to User object
- Hydrates User object with contacts
- Removes contacts based on permission in formatter
- Allows for searching of users by contact

Test checklist:
- [ ]

- [ ] I certify that I ran my checklist

Fixes ushahidi/platform# .

Ping @ushahidi/platform

### CODE REVIEW

**Comment by willdoran** [comment]
If I have Post entity derived from a data source, if the Message the Post is derived from is associated with the Contact of an existing User. That User should be set as the user id of the Post.

For existing data this may require a migration akin to 

update messages set user_id = ( select user_id from contacts c where messages.contact_id = c.id) where user_id is null;
update posts set user_id = ( select user_id from messages m where m.post_id = posts.id) where user_id is null;

File: application/classes/Ushahidi/Repository/User.php [hunk 1]
```diff
@@ -54,9 +54,28 @@ protected function getTable()
...
+		if (!empty($data['id']))
+		{
+			$data += [
+				'contacts' => $this->getContacts($data['id']),
+			];
+		}
 		return new User($data);
 	}
 
+	protected function getContacts($entity_id)
```
Comment (rjmackay) [inline application/classes/Ushahidi/Repository/User.php]: Hopefully we can eventually just combine contacts + users so this gets less circular

File: src/Core/Entity/User.php [hunk 2]
```diff
@@ -27,6 +27,7 @@ class User extends StaticEntity
 	protected $updated;
 	protected $role;
 	protected $language;
+	protected $contacts;
```
Comment (rjmackay) [inline src/Core/Entity/User.php]: Could you check what happens if someone sends an update with contacts in the payload?
We've seen previous issues where the repo's try to automatically set a user.contacts field in the db and everything breaks.

Comment (willdoran) [reply src/Core/Entity/User.php]: I tested it via the client and it worked but I can add a test for it

Comment (willdoran) [reply src/Core/Entity/User.php]: Ok, I updated one of the tests to ensure that including contacts doesn't cause a problem and changed the user update method to unset contacts explicitly.

File: application/classes/Ushahidi/Repository/User.php [hunk 3]
```diff
@@ -114,13 +134,18 @@ public function getSearchFields()
...
 		if ($search->q)
 		{
 			$query->and_where_open();
 			$query->where('email', 'LIKE', "%" . $search->q . "%");
 			$query->or_where('realname', 'LIKE', "%" . $search->q . "%");
 			$query->and_where_close();
+
+			// Adding search contacts
+			$query->join('contacts')->on("$table.id", '=', 'contacts.user_id')
+			->or_where('contacts.contact', 'like', '%' . $search->q . '%');
```
Comment (rjmackay) [inline application/classes/Ushahidi/Repository/User.php]: @willdoran it occurs to me now that we should probably limit searching by email or contact to users with permissions, potentially you could ID a user via search despite the fact the API doesn't return their email or contact info.

Comment (willdoran) [reply application/classes/Ushahidi/Repository/User.php]: True we should add a check, I'll patch it later today.

**Pull request merged**